tools/mdm/migration/mdmproxy/mdmproxy.go

```diff
@@ -0,0 +1,312 @@
+package main
+
+import (
+	"bufio"
+	"bytes"
+	"errors"
+	"flag"
+	"fmt"
+	"hash/fnv"
+	"io"
+	"log"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"howett.net/plist"
+)
+
+type mdmProxy struct {
+	migrateUDIDs      map[string]struct{}
+	migratePercentage int
+	existingServerURL string
+	existingHostname  string
+	fleetServerURL    string
+	existingProxy     *httputil.ReverseProxy
+	fleetProxy        *httputil.ReverseProxy
+	// mutex is used to sync reads/updates to the migrateUDIDs and migratePercentage
+	mutex sync.RWMutex
+	// token is used to authenticate updates to the migrateUDIDs and migratePercentage
+	token string
+}
+
+func (m *mdmProxy) handleProxy(w http.ResponseWriter, r *http.Request) {
+	if r.URL.Host != "" {
+		log.Printf("%s %s Forbidden", r.Method, r.URL.String())
+		http.Error(w, "Forbidden", http.StatusForbidden)
+		return
+	}
+
+	// Send all SCEP requests to the existing server
+	if strings.Contains(r.URL.Path, "scep") {
+		log.Printf("%s %s -> Existing (SCEP)", r.Method, r.URL.String())
+		m.existingProxy.ServeHTTP(w, r)
+		return
+	}
+
+	// Send all micromdm API requests to the existing server
+	if strings.HasPrefix(r.URL.Path, "/v1") || strings.HasPrefix(r.URL.Path, "/push") {
+		log.Printf("%s %s -> Existing (API)", r.Method, r.URL.String())
+		m.existingProxy.ServeHTTP(w, r)
+		return
+
+	}
+
+	// Read the body of the request
+	body, err := io.ReadAll(r.Body)
+	_ = r.Body.Close()
+	if err != nil {
+		log.Println("Failed to read request body: ", err.Error())
+		http.Error(w, "Unable to read request body", http.StatusUnprocessableEntity)
+		return
+	}
+	// Reset body so that the reverse proxy request includes it
+	r.Body = io.NopCloser(bytes.NewReader(body))
+
+	// Get the UDID from request
+	udid, err := udidFromRequestBody(body)
+	if err != nil {
+		log.Printf("%s %s Failed to get UDID: %v", r.Method, r.URL.String(), err)
+		http.Error(w, err.Error(), http.StatusUnprocessableEntity)
+		return
+	}
+
+	// Migrated UDIDs go to the Fleet server, otherwise requests go to the existing server.
+	if udid != "" && m.isUDIDMigrated(udid) {
+		log.Printf("%s %s (%s) -> Fleet", r.Method, r.URL.String(), udid)
+		m.fleetProxy.ServeHTTP(w, r)
+	} else {
+		log.Printf("%s %s (%s) -> Existing", r.Method, r.URL.String(), udid)
+		m.existingProxy.ServeHTTP(w, r)
+	}
+}
+
+func (m *mdmProxy) handleUpdatePercentage(w http.ResponseWriter, r *http.Request) {
+	if m.token == "" {
+		http.Error(w, "Set auth token to enable remote updates", http.StatusUnauthorized)
+		return
+	}
+	token := r.Header.Get("Authorization")
+	if token == "" {
+		http.Error(w, "Authorization header must be provided", http.StatusUnauthorized)
+		return
+
+	}
+	if token != m.token {
```

Comment:
I don't think this is super important, but might be unexpected for consumers so just throwing the heads up: the expected header here is `Authorization: <token>`, it doesn't contain the type (eg: `Authorization: Bearer <token>`)

Reply:
Yeah I think it makes sense to be consistent with Fleet server conventions.